Client Library HIL | query calls

SahilTikale opened the pull request:
Current PR, includes the basic code structure required to built a client library for HIL.
It also includes client library support for all query calls.

Still pending:
Unit tests for the new code.
Making correspondng cli calls to go via Client library.

zenhack commented:
@SahilTikale, you'll need to work in the changes from #618.

Also, could you go over this for style issues? Lots of PEP8/PEP257 violations here. *please* make sure you've read those documents and understood them, and follow the guidelines. The codebase is full of deviations in general, but let's not make it worse.

Additionally, if you're using vim, I reccomend python-mode, which will warn you about a lot of these issues, plus things like unused imports, and will add other nice python-aware stuff to your editor as well:

https://github.com/klen/python-mode

@gsilvis, @henn, can we prioritize the pep8 testing soonish?

shwsun commented:
I would point out that Emacs users could adopt [py-yapf](https://github.com/paetzke/py-yapf.el) and [YAPF](https://github.com/google/yapf) is also interesting to look into.

zenhack commented:
yapf looks interesting; my experience with gofmt has been excellent, it would be great to have something like that for us. It's says the thing is still fairly immature though, so I'm worried about upgrading versions and then have my editor accidentally change a bunch of formatting without me having actually made any meaningful changes to the file. Even worse would be if two developers were using different versions of the tool, and kept thrashing, making surrious changes to the source...

I gave it a try and it made a few odd choices, e.g.

```diff
-        return model.NetworkAttachment.query.filter(
-            model.NetworkAttachment.nic == nic,
-            query,
-        ).count() != 0
+        return model.NetworkAttachment.query.filter(model.NetworkAttachment.nic
+                                                    == nic,
+                                                    query, ).count() != 0
```

...but nothing I think not worth tolerating to just not spend review time on formatting issues at all.

I'd want to verify we can rely on consistent output from release to release, but given that I love the idea of instituting this.

henn commented on haas/client/node.py:
```diff
@@ -0,0 +1,33 @@
+        q = requests.get(url, headers={"Authorization": "Basic %s" %self.auth})
+        return q.json()
+
+
```
blank lines

SahilTikale replied:
fixed it.

henn commented on haas/cli.py:
```diff
@@ -27,6 +27,21 @@
+from haas.client.client import Client
+
+
+ep = "http://127.0.0.1:5000" or os.environ.get('HAAS_ENDPOINT')
```
a) This should be in the client library, and
b) we shouldn't have defaults for these. That could cause unexpected surprises!

SahilTikale replied:
We need a way to choose an authentication system for the client library. CLI is the interface via which we can choose appropriate auth from the client library.

henn commented on haas/client/auth.py:
```diff
@@ -0,0 +1,13 @@
+import base64
+
```
lots of blank lines

SahilTikale replied:
fixed, passes pep8

henn commented on haas/client/auth.py:
```diff
@@ -0,0 +1,13 @@
+
+
+
+def auth_db(username, password):
```
To avoid reinvention, after #615 merges I would suggest copying the various HTTPClient() implementations from cli.py. Eventually, when cli.py is completely calling your client library, we can then remove the original HTTPClient().

zenhack replied:
There's no need to wait for #615; the `HTTPClient` stuff is already merged (via #626).

SahilTikale replied:
Done. moved into client library.

henn commented on haas/client/client.py:
```diff
@@ -0,0 +1,19 @@
+from haas.client.node import Node
+from haas.client.project import Project
+from haas.client.switch import Switch
+#from haas.client.network import Network
```
If we don't need these commented out lines, maybe just remove them?

SahilTikale replied:
It was reminder for future work. It is done now. fixed.

henn commented on haas/client/base.py:
```diff
@@ -0,0 +1,41 @@
+    """
+
+
+    def __init__(self, endpoint=None, auth=None):
```
If `endpoint` and `auth` are required, could we remove the default values and check later on?

SahilTikale replied:
The default is there so as to get the setup as easy as possible to start with. 
Default being no authentication required.

henn commented on haas/client/client.py:
```diff
@@ -0,0 +1,19 @@
+from haas.client.switch import Switch
+#from haas.client.network import Network
+from haas.client.auth import auth_db
+
```
blank lines

SahilTikale replied:
fixed.

henn commented on haas/client/node.py:
```diff
@@ -0,0 +1,33 @@
+    """ Consists of calls to query and manipulate node related
+    objects and relations """
+
+    def list(self, is_free):
```
Should is_free have a default (like maybe `False`)?

SahilTikale replied:
I checked the cli, client and api code for this call. CLI rejects any other options other than `free` and `all`. But api and client library do not have any such restrictions. I mean in absence of the check done at CLI, the API would respond to any string as if it was keyword `all`. I believe this should be fixed at API level.

henn commented on haas/client/node.py:
```diff
@@ -0,0 +1,33 @@
+    objects and relations """
+
+    def list(self, is_free):
+        """ List all nodes that HIL manages """
```
We need to describe what `is_free` does in here

SahilTikale replied:
explained in the doc-string. you will see in the next commit.

henn commented on haas/client/node.py:
```diff
@@ -0,0 +1,33 @@
+    def list(self, is_free):
+        """ List all nodes that HIL manages """
+        url = self.object_url('node', is_free)
+        q = requests.get(url, headers={"Authorization": "Basic %s" % self.auth})
```
If we use the HTTPClient implementation, we don't need to do basic auth here.

Also, requests allows you to [pass an `auth` parameter](http://docs.python-requests.org/en/master/user/authentication/) so you don't have an manually insert a header line.

SahilTikale replied:
removed that when adding support for keystone auth.

henn commented on haas/client/node.py:
```diff
@@ -0,0 +1,33 @@
+        if q.ok:
+            return q.json()
+        elif q.status_code == 401:
+            raise AuthenticationError("Make sure credentials match chosen authentication backend.")
```
What about other statuses (404, 400, 500, etc)?

SahilTikale replied:
Could not find cases that would raise those errors. If you can help me with that, will include those.

henn commented on haas/client/node.py:
```diff
@@ -0,0 +1,33 @@
+        self.node_name = node_name
+        url = self.object_url('node', node_name)
+        q = requests.get(url, headers={"Authorization": "Basic %s" %self.auth})
+        return q.json()
```
Should we be checking for error status?

It might make sense to factor that out (it's going to happen for everything).

henn commented on haas/client/project.py:
```diff
@@ -0,0 +1,43 @@
+
+
+        def networks_in(self, project_name):
+            """ Lists nodes allocated to project <project_name> """
```
Think you meant `Lists networks...`

SahilTikale replied:
No. I meant networks in a project. I wanted it to look like project.networks_in(<project_name>) as in "give me networks that belong to < project name>"

henn commented on haas/client/project.py:
```diff
@@ -0,0 +1,43 @@
+            elif q.status_code == 401:
+                raise AuthenticationError("Make sure credentials match chosen authentication backend.")
+
+        def nodes_in(self, project_name):
```
Could we rename this and the next one to nodes() and networks() ?

henn commented on haas/client/project.py:
```diff
@@ -0,0 +1,43 @@
+
+        def nodes_in(self, project_name):
+            """ Lists nodes allocated to project <project_name> """
+            self.project_name = project_name
```
Why assign this locally? Will future calls of the sort `Project.operation()` assume a built-in project?

If that's the case, maybe have an `__init__()` that binds the object to a particular class, then make this argument optional except when someone's calling the class vs an instantiated object?

henn commented on haas/client_lib/client_lib.py:
```diff
@@ -0,0 +1,91 @@
+
+
+## ** QUERY COMMANDS **
+    def list_free_nodes(self):
```
I think you already covered this functionality in Node.list().

henn commented:
I guess we need to decide whether we want to stick with the object-oriented version API or have a single class instance that has a flat namespace; currently both are implemented.

I could go either way; we just need to pick one and be consistent so that folks know where to find functionality. 

@knikolla - I know you have strong opinions on things like this.

henn commented on tests/unit/client.py:
```diff
@@ -0,0 +1,301 @@
+            '[headnode]',
+            'base_imgs = base-headnode, img1, img2, img3, img4',
+            '[database]',
+            'uri = sqlite:///%s/haas.db' % tmpdir,
```
Why not use in-memory DB?

zenhack replied:
I don't think that would work, since you call the cli just to initialize the db, and then again to start the server. If you used an in-memory db you'd lose it when the first process exits.

henn commented on tests/unit/client.py:
```diff
@@ -0,0 +1,301 @@
+    os.chdir(tmpdir)
+    with open('haas.cfg', 'w') as f:
+        config = '\n'.join([
+            '[general]'
```
Need trailing comma

zenhack commented on tests/unit/client.py:
```diff
@@ -245,7 +245,7 @@ def create_setup(request):
     proc = run_server(['haas', 'serve', '8888' ])
     import time
-    time.sleep(0.5)
+    time.sleep(1)
```
You still have a race condition, even after this change. In theory this will be true no matter what the parameter to sleep is, and a truly "safe" value is probably going to make our ci run longer than we'd like.

There's a fixture in `tests/integration/keystone.py` on #615 that does this correctly. That should be merged relatively soon, maybe when it is you can pull it in, abstract out that logic and use it here as well?

shwsun commented:
Maybe the integration test errors will go away as well if you merge the 11 commits behind.

zenhack commented:
@SahilTikale, would you merge master back into this? Also, #628 just landed, so you'll *need* to fix those pep8 errors now.

SahilTikale commented:
@zenhack @shwsun 
Will do that at the earliest when I have breathing space from NSDI.
Things seem to be in control here so will take a stab at this.

naved001 commented:
Hi @SahilTikale have you had time to look at the requested changes for this pr to move? Thanks!

zenhack commented on haas/client/node.py:
```diff
@@ -0,0 +1,121 @@
+        elif q.status_code == 401:
+            raise errors.AuthenticationError(
+                  "Make sure credentials match chosen authentication backend."
+                  )
```
There's too much boilerplate/repetitive code re: error handling both here and in the CLI module; find a way to cut down on it. Note that the bodies of the responses in the errors the server returns can probably be leveraged here.

SahilTikale replied:
Cli is going to get redesigned so not worrying about it too much. As to the boiler plate code like above I agree about that. some errors look are common for all classes and should be inherited from the parent class. I do not have a good way of implementing it yet so left it like this for now. With recent additions to api and new api in the pipeline, I think I will let it be and refactor it at a later time when client library is relatively stable.

zenhack commented on haas/client/node.py:
```diff
@@ -0,0 +1,121 @@
+                    )
+        elif q.status_code == 500:
+            raise errors.NotFoundError(
+                    "Operation Failed. Contact your system administrator"
```
Probably should be more explicit that this is a server-side problem.

SahilTikale replied:
Noted. Fixed it.

zenhack commented on haas/client/node.py:
```diff
@@ -0,0 +1,121 @@
+                    )
+
+    def add_nic(self, node_name, nic_name, macaddr):
+        """ adds a <nic> from <node>"""
```
s/from/to/

SahilTikale replied:
oops. fixed it.

zenhack commented on haas/client/node.py:
```diff
@@ -0,0 +1,121 @@
+    def show_node(self, node_name):
+        """ Shows attributes of a given node """
+
+        self.node_name = node_name
```
I don't think you should be adding attributes to `self` here; the `Node` class is just a namespace for methods that call the server, right? it's not a single node object.

SahilTikale replied:
this is fixed.

zenhack commented on haas/client/node.py:
```diff
@@ -0,0 +1,121 @@
+
+    def disconnect_network(self, node_name, nic, network):
+        """ Disconnect <node> from <network> on the given <nic>. """
+        pass
```
Still need implementations for these.

SahilTikale replied:
for this PR, I do not wish to add more calls. Each calls means changing equivalent code in CLI and writing tests. In future I will make smaller PR requests for these and other missing calls.

zenhack commented on haas/client/project.py:
```diff
@@ -0,0 +1,107 @@
+            elif q.status_code == 404:
+                raise errors.NotFoundError("Node or project does not exist.")
+
+        def disconnect(self, project_name, node_name):
```
Disconnect what? Perhaps call this `detach_node`, like in `rest_api.md`? This goes for most things -- keep the names consistent and descriptive.

SahilTikale replied:
as per #562 I am trying to make some operation names consistent. If noone likes it, will change it back.

zenhack replied:
One change at a time; this pr is already enormous, please keep things on-topic. Changing the names should be it's own pr.

SahilTikale replied:
ok, changed it back from `disconnect` to `detach`

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,442 @@
+import haas
+from haas import model, api, deferred, server, config
+from haas.model import db
+from haas.test_common import *
```
Don't do `import *`. I know it's used elsewhere, but we want to remove that. Let's not make it worse. See also #678.

SahilTikale replied:
I agree. Sorry about that. Fixed it.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,442 @@
+from urlparse import urljoin
+import requests
+from requests.exceptions import ConnectionError
+# Hook to the client library
```
I'm not sure what this comment is supposed to mean or why it's here. Same goes for the one in the cli.

SahilTikale replied:
That was to help me focus on the new code. Removed it.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,442 @@
+        return (tmpdir, cwd)
+
+
+def cleanup((tmpdir, cwd)):
```
Why is there an extra set of parens here?

SahilTikale replied:
they were symbolic of the dark circles I got while doing this PR for several late nights or early mornings depending on how you wanna see it. :)
Removed it.

SahilTikale replied:
I do not remember now, but I guess I borrowed it from some python documentation about unit tests.
without which all my unit tests were failing.

SahilTikale replied:
`cleanup` functions takes a tuple as an input that is why it looks like it has two parenthesis. `make_config` returns a tuple which is `cleanup` function uses to clean up after all tests are done.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,442 @@
+    tests finish.
+    """
+    proc = Popen(cmd)
+    return proc
```
This function is literally just `Popen`.

The keystone ci tests arlready have some fixtures for running a real haas server; these should probably be moved to test_common and used here as well. We don't need to use the CLI to launch stuff here since it's a test of the client library, but we may want to elsewhere do some basic tests that the server responds when launched via the CLI. Unit tests should be as self-contained as possible though.

SahilTikale replied:
I think we should do a separate PR for tests infrastructure related changes.

zenhack replied:
I'm fine not doing big infrastructure changes, but this function is still literally just `Popen`; I'd delete it and change the calls to it to direct calls to Popen.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,442 @@
+            'type': api_name+'brocade', 'hostname': 'brocade-01',
+            'username': 'root', 'password': 'root1234',
+            'interface_type': 'TenGigabitEthernet'
+            }
```
You could make this a list and then turn the function calls below into a for loop.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,442 @@
+    requests.put(url_node+'node-05', data=json.dumps({"obm": obminfo5}))
+    requests.put(url_node+'node-06', data=json.dumps({"obm": obminfo6}))
+    requests.put(url_node+'node-07', data=json.dumps({"obm": obminfo7}))
+    requests.put(url_node+'node-08', data=json.dumps({"obm": obminfo8}))
```
You should be able to colapse everything from the start of the function to here into a for loop.

SahilTikale replied:
I did try collapsing it, but there is unique info that would be too convulated to be looped for `obminfoX` hence chose verbose clarity over confused tersity.

zenhack replied:
If you make obminfoX a list instead of independent varaibles, you can just do:

```python
for i in range(obminfo):
    requests.put(url_node+('node-%02d' % (i+1)), data=json.dumps(obminfio[i]))
```

SahilTikale replied:
fixed it in one of the previous commits.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,442 @@
+# software distributed under the License is distributed on an "AS
+# IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied.  See the License for the specific language
+# governing permissions and limitations under the License.
```
So I **really** like some of the fixtures here, and I think there a great basis for writing some much needed tests of the server -- there's nowhere else where we're really testing the HTTP bits of the API; we're mostly calling the functions directly. However, it probably doesn't belong in the client library unit tests.

Furthermore, I don't think you need to run a real server process for these; you could just implement a mock HTTPClient that uses flask's test client.

I'd like to simplify this module a bit; it's a lot to digest, and much of it is inessential to unit testing the client library. That said, I *don't* want to just junk what you've written here, since it seems like a really good bit of integration testing for the server.

SahilTikale replied:
Hey @zenhack, 
This was settled in a meeting. @pjd-nu @okrieg both agree that it is a good idea to have a running server as it setup a good way to test client library externally.

zenhack replied:
There's still some duplication between this and stuff I remember writing for the keystone backend, which I want sorted out before we merge. I think I agreed to send you a PR for that bit though; will do that soon.

zenhack commented:
@SahilTikale, I took a look at trying to do some testing re: keystone, but this is pretty far behind master and there are a few things there that would be nice to have. Could you try merging master back in/rebasing? the merge conflicts are only going to get worse the longer you wait.

zenhack commented:
@SahilTikale, I would *really* encourage you to merge master back into this before too long; it's just going to get harder and harder to deal with the conflicts.

SahilTikale commented:
I am trying to complete the CLI interfacing with client library. Once I have a working code. 
I will merge it with master and work at resolving the conflicts.

SahilTikale commented on haas/cli.py:
```diff
@@ -721,15 +766,18 @@ def list_networks():
-    url = object_url('node', node)
-    do_get(url)
+    """Display information about a <node>
+    FIXME: Recursion should be implemented to the output.
```
The output of show_node is a dictionary that can be list of list, having multiple nics and networks. 
More meta data about node could be shown via this call. So this is a note for future developers who will be making the CLI to make appropriate changes.

zenhack replied:
Could you modify the comment to include that clarification?

SahilTikale commented on haas/client/auth.py:
```diff
@@ -0,0 +1,17 @@
+
+def db_auth(username, password):
+    """ For database as authentication backend, this function prepares
+    the default http client session using requests module.
```
For the same reason (no. 2) that you mentioned I wanted to call it db_auth, so distinguish it from other authentication back-ends. This one clarifies that this back-end resides in the database.

SahilTikale replied:
Point 1. Addressed. Made the docstring terse.

SahilTikale commented on haas/client/base.py:
```diff
@@ -0,0 +1,38 @@
+        """ Initialize an instance of the library with following parameters.
+        endpoint: stands for the http endpoint eg. endpoint=http://127.0.0.1
+        user: username as which you wish to connect to HaaS
+        password: password for the 'user' as decribed above.
```
Noted and addressed. Next commit has the changes.

SahilTikale commented on haas/client/base.py:
```diff
@@ -0,0 +1,38 @@
+        if None in [self.endpoint, self.s]:
+            raise LookupError(
+                    "Insufficient attributes to establish "
+                    "connection with HaaS server")
```
This message gets raised when people start accessing HIL without initializing the client library. That is not having the requisite environment variable. Without the LookupError, it throws lot of lines on the output and user is left to figure out what went wrong. I felt this is going to be very common issue with new people getting acquainted with HIL, hence a good error messsage may save some time. As to its opaque nature, I will try to make it more informative.

SahilTikale commented on haas/client/network.py:
```diff
@@ -0,0 +1,21 @@
+
+
+class Network(ClientBase):
+        """ Consists of calls to query and manipulate project related
```
Good catch. Fixed it.

SahilTikale commented:
merged with master. and addressed all the comments from @zenhack.
will add documentation soon. This is ready for review @shwsun @henn @naved001

zenhack commented:
So, this is a *big* pull request. I think a lot is somewhat tangential to client library itself, but in many cases still work worth not throwing away. What I'd like to see is for the conceptually self contained changes to get split off into standalone pull requests, to make them easier to review.

A prime candidate for this is the changing HTTP status codes; it's really an orthoganal change to the client lib, but I absolutely agree that we're overusing 409 and BlockedError in particular. @SahilTikale, could you do some cherry-picking and get me a PR that just has the status code changes on top of master? Then we'll rebase this one on top of it, and go from there.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,573 @@
+    url_node = 'http://127.0.0.1:8888/node/'
+    api_nodename = 'http://schema.massopencloud.org/haas/v0/obm/'
+
+    obminfo1 = {
```
Just use a list, instead of obminfo1, obminfo2...

SahilTikale replied:
this one was addressed in one of the previous commits.

henn commented:
@SahilTikale - would you mind addressing the test failure when you have a chance and updating? It looks like [it is the result](https://travis-ci.org/CCI-MOC/hil/jobs/186197982) of some PEP8 issues and test_add_list_delete_projects() in tests/integration/cli.py.

zenhack commented:
Quoting Sahil Tikale (2017-01-31 19:18:24)

>    cleanup functions takes a tuple as an input that is why it looks like
>    it has two parenthesis. make_config returns a tuple which is cleanup
>    function uses to clean up after all tests are done.

Makes sense, cool.

SahilTikale commented:
```
(.venv)sahil@rHaaS:~/git/HIL$ py.test tests/unit/
============================================== test session starts ===============================================
platform linux2 -- Python 2.7.6, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: /home/sahil/git/HIL, inifile: setup.cfg
plugins: xdist-1.14, cov-1.8.0
collected 543 items 

tests/unit/class_resolver.py ...
tests/unit/cli.py ...
tests/unit/client.py ........................................
tests/unit/config.py .
tests/unit/dev_support.py ....
tests/unit/haas_auth.py .
tests/unit/migrations.py s
tests/unit/model.py ..............
tests/unit/old_client.py ........................................
tests/unit/rest.py ....................................
tests/unit/test_common.py ...................................
tests/unit/api/auth.py .........................................................................................................................................................................
tests/unit/api/main.py ...........................................................................................................................................................
tests/unit/ext/auth/database.py ...............................
tests/unit/ext/auth/mock.py .....
tests/unit/ext/network_allocators/vlan_pool.py ..
tests/unit/ext/switches/brocade.py ...

==================================== 542 passed, 1 skipped in 239.77 seconds =====================================
```

It is passing all unit tests in my development environment. I wonder why `tests/unit/client.py::Test_Node::test_show_node`  test is failing here.

henn commented:
It looks like we need to include `keystoneauth1` now in setup.py, due to getting this error:

```
In [2]: from haas.client import client
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-2-eabf120236c7> in <module>()
----> 1 from haas.client import client

/home/haas/hil2/haas/client/client.py in <module>()
      6 from haas.client.network import Network
      7 from haas.client.user import User
----> 8 from haas.client import auth
      9 
     10 

/home/haas/hil2/haas/client/auth.py in <module>()
      1 import requests
----> 2 import keystoneauth1.identity
      3 from keystoneauth1.identity import v3
      4 from keystoneauth1 import session
      5 

ImportError: No module named keystoneauth1.identity
```

zenhack commented:
I'm really going to insist on this: Split up the PR. I am not interested in reviewing a patch that is this big and has so many different changes in it. I **will not merge this** as a single PR; there's no way we can do a quality code review of a patch this size, and there's no reason it should be this size.

SahilTikale commented:
I think we should talk @zenhack.

zenhack commented:
I agree. I'm fighting a cold right now, but sometime soon I'm going to want to come in to work on drivers re: port_revert. I'll send you an email and we can pick a time when you'll be around so we can chat.

SahilTikale commented:
replying to 

> So, this is a big pull request. I think a lot is somewhat tangential to client library itself, but in many cases still work worth not throwing away. What I'd like to see is for the conceptually self contained changes to get split off into standalone pull requests, to make them easier to review.
A prime candidate for this is the changing HTTP status codes; it's really an orthoganal change to the client lib, but I absolutely agree that we're overusing 409 and BlockedError in particular. @SahilTikale, could you do some cherry-picking and get me a PR that just has the status code changes on top of master? Then we'll rebase this one on top of it, and go from there.

I will make a new PR, adding the HTTP status codes. then I will delete it from the current PR, or leave it as it is and we will see when we split the PR further. 

Sorry for delayed response.

SahilTikale commented:
@zenhack let me know what you think about this refactor ? Added a new function in `client/base.py` that is reused in `client/project.py`. Similarly added a function `check_clientlib_response` in `cli.py` and modified project related calls to use it.  
Once you are satisfied with the quality of refactor, I will implement it for rest of the PR.

SahilTikale commented:
All my tests are passing:
```
(.venv)sahil@rHaaS:~/git/HIL$ py.test tests/unit/
===================================================================== test session starts =====================================================================
platform linux2 -- Python 2.7.6, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: /home/sahil/git/HIL, inifile: setup.cfg
plugins: xdist-1.14, cov-1.8.0
collected 503 items 

tests/unit/class_resolver.py ...
tests/unit/cli.py ...
tests/unit/client.py ........................................
tests/unit/config.py .
tests/unit/dev_support.py ....
tests/unit/haas_auth.py .
tests/unit/migrations.py s
tests/unit/model.py ..............
tests/unit/rest.py ....................................
tests/unit/test_common.py ...................................
tests/unit/api/auth.py .........................................................................................................................................................................
tests/unit/api/main.py ...........................................................................................................................................................
tests/unit/ext/auth/database.py ...............................
tests/unit/ext/auth/mock.py .....
tests/unit/ext/network_allocators/vlan_pool.py ..
tests/unit/ext/switches/brocade.py ...

=========================================================== 502 passed, 1 skipped in 153.09 seconds ===========================================================(.venv)sahil@rHaaS:~/git/HIL$ py.test tests/unit/
===================================================================== test session starts =====================================================================
platform linux2 -- Python 2.7.6, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: /home/sahil/git/HIL, inifile: setup.cfg
plugins: xdist-1.14, cov-1.8.0
collected 503 items 

tests/unit/class_resolver.py ...
tests/unit/cli.py ...
tests/unit/client.py ........................................
tests/unit/config.py .
tests/unit/dev_support.py ....
tests/unit/haas_auth.py .
tests/unit/migrations.py s
tests/unit/model.py ..............
tests/unit/rest.py ....................................
tests/unit/test_common.py ...................................
tests/unit/api/auth.py .........................................................................................................................................................................
tests/unit/api/main.py ...........................................................................................................................................................
tests/unit/ext/auth/database.py ...............................
tests/unit/ext/auth/mock.py .....
tests/unit/ext/network_allocators/vlan_pool.py ..
tests/unit/ext/switches/brocade.py ...

=========================================================== 502 passed, 1 skipped in 153.09 seconds ===========================================================
```

zenhack commented on haas/client/project.py:
```diff
@@ -0,0 +1,64 @@
+
+        def nodes_in(self, project_name):
+            """ Lists nodes allocated to project <project_name> """
+            self.project_name = project_name
```
Why are you assigning this? It shouldn't need tou outlive the function call?

zenhack commented:
@SahilTikale, I like the refactoring you've done re: error handling; I think it makes sense to apply it to the rest of the calls.

SahilTikale commented:
Hi @zenhack made all the changes as requested. Have a look and let me know.

zenhack commented on haas/cli.py:
```diff
@@ -179,13 +195,16 @@ def setup_http_client():
 
     This may be extended with other backends in the future.
     """
-    global http_client
```
@SahilTikale, you still haven't addressed this (though the comment has been hidden because you changed code nearby).

There are a number of other comments that I made in the past that you've not addressed -- please pick through them and do so.

SahilTikale replied:
This issue is addressed long back. Now the CLI uses this variable for backward compatibility since client library does not support all calls yet (node_register, switch_register and headnode related calls).

SahilTikale commented on haas/cli.py:
```diff
@@ -27,6 +27,22 @@
+
+#sess = db_auth(username, password)
+
+#C = Client(ep, sess) #Initializing client library
```
This is fixed.

SahilTikale commented on haas/cli.py:
```diff
@@ -684,24 +728,25 @@ def list_nodes(is_free):
-    url = object_url('nodes', is_free)
-    do_get(url)
-
+    q = C.node.list(is_free)
```
@zenhack I find it more appropriate to make it Boolean. What do you think ? Like if `is_free` is `True` show only free nodes else show all. But that needs to be changed in API call too. I am not sure if that is out of the context of the PR.  Let me know what you think.

SahilTikale commented on haas/cli.py:
```diff
@@ -684,24 +728,25 @@ def list_nodes(is_free):
+    elif is_free == 'free':
+        sys.stdout.write('Free nodes {}\t:    {}\n'.format(len(q), " ".join(q)))
+    else:
+        sys.stdout.write('Error: {} is an invalid argument\n'.format(is_free))
```
fixed these. next commit will have the changes.

SahilTikale commented:
@zenhack I have addressed all of your comments in this commit. Let me know if I missed any.

zenhack commented on setup.py:
```diff
@@ -96,4 +96,5 @@ def _get_readme():
                         'pep8>=1.7.0',
                         'requests_mock>=1.0.0,<2.0',
                         'lxml>=3.6.0,<4.0',
+                        'keystoneauth1>=2.16,<3.0',
```
We specifically *don't* want this in setup.py; that makes it a mandatory requirement, even if the user isn't planning on using the keystone auth extension.

SahilTikale replied:
I will remove this once we figure out the refactor as suggested in your later comments (removing `auth.py`)

zenhack commented on haas/client/base.py:
```diff
@@ -0,0 +1,50 @@
+    appropriate message.
+    """
+
+    def __init__(self, endpoint=None, sess=None):
```
Below you explicitly check if these are None and throw an error; if these are mandatory just don't supply the defaults.

SahilTikale replied:
I did not understand your comment about `if these are mandatory just don't supply the defaults.`
I am checking if they are explicitly `None` and raises error in that case.

zenhack commented on haas/client/errors.py:
```diff
@@ -0,0 +1,22 @@
+    pass
+
+
+class ProjectMismatchError(Exception):
```
Now that you've refactored things to de-duplicate the error checking, I don't think you're using any of the errors in this file except for FailedAPICallException. In that case, you should get rid of the others, and probably move the remaining one to another module, just because it seems silly to have a whole file just for the one trivial error type.

SahilTikale replied:
I wanted to keep the client library confined to its folder. I am not importing from the `haas` folder anything. Is it too bad to let this error.py stay with just one exception defined in it ?

zenhack commented on haas/client/node.py:
```diff
@@ -0,0 +1,93 @@
+#       FIXME: In future obm_types should be dynamically fetched
+#        from haas.cfg, need a new api call for querying available
+#        and currently active drivers for HIL
+        pass
```
This method doesn't currently do anything. I know you said something about not actually filling out every method, but if we're going to do that I want the ones we don't implement to be marked as not implemented.

SahilTikale replied:
Should I write `Not implemented` in the doc string or as a comment.

zenhack commented on haas/client/switch.py:
```diff
@@ -0,0 +1,58 @@
+#       HIL documentation to use appropriate flags to register
+#       it with HIL.
+
+        pass
```
You don't need pass if you've got a docstring; please go through and remove it in cases like these where it isn't necessary.

SahilTikale replied:
will fix that in the next commit.

zenhack commented on haas/client/user.py:
```diff
@@ -0,0 +1,46 @@
+        """Create a user <username> with password <password>.
+    <privilege> may by either "admin" or "regular", and determines whether a
+    user is authorized for administrative privileges
+    """
```
1. Blank line between the one-line summary and the rest of the docstring, per pep257.
2. indent the rest of the docstring to match the first line.

(1) applies to many ohter docstrings in this patch; please go through and fix it everywhere.

SahilTikale replied:
Ok. will fix that. Strangely the pep8 tests did not catch this.

zenhack commented on haas/client/user.py:
```diff
@@ -0,0 +1,46 @@
+        url = self.object_url('/auth/basic/user', username)
+
+        if privilege not in('admin', 'regular'):
+            raise TypeError(
```
Would suggest ValueError here; TypeError would make more sense if e.g. you passed in an integer.

SahilTikale replied:
Noted. next commit fixes it.

zenhack commented on haas/client/user.py:
```diff
@@ -0,0 +1,46 @@
+        # Note: Named such so that in future we can have granular
+        # access to projects
+        # like "grant_read_access", "grant_write_access",
+        # "grant_all_access", etc
```
I'd rather have this mirror the names we use server side (or, more importantly in rest_api.md).

SahilTikale replied:
Made them simple as `add` and `remove` because you can only add or remove user to projects

zenhack commented on tests/integration/cli.py:
```diff
@@ -15,21 +15,24 @@ def test_add_list_delete_projects():
     output = subprocess.check_output(['haas', 'list_projects'])
-    projects = json.loads(output)
+    projects = output.split(" ")
+    projects = map(str.strip, projects)  # strips off \n from project name
```
Might suggest `map(lambda s: s.strip(), projects)` here. I know the json library actually hands back unicode objects, not str, and that would cause an error. This is actually fine since it comes from subprocess.check_output, which returns str, but I had to stare at it a bit. The lambda is more obviously correct regardless of which type the object is. You could also use a list comprehension:

`projects = [p.strip() for p in projects]`

Which I actually like a bit better than the lambda.

SahilTikale replied:
I did not make these changes. I do not know how they crept into my commit. Travis is failing the exact same tests for my PR as of #729

naved001 replied:
Again, this was me. sorry for the confusion. that was me trying to fix the failing test.

naved001 replied:
also, we could do  `projects.remove('\n')` since we want to remove `\n` here. I don't remember why I used map.

zenhack replied:
Any of the above solutions are ok with me, but please implement one of them.

zenhack commented on tests/integration/cli.py:
```diff
@@ -15,21 +15,24 @@ def test_add_list_delete_projects():
     output = subprocess.check_output(['haas', 'list_projects'])
-    projects = json.loads(output)
+    projects = output.split(" ")
+    projects.remove('\n')  # remove \n from list
```
This comment is just noise, I would remove it -- it's just describing what the remove method does, and doesn't really provide any insight into our own code.

SahilTikale replied:
I did not make this change. or to any tests other than unit tests for my library.

naved001 replied:
This was a commit I had pushed to Sahil's branch while he was away. This test was failing and I had added that comment/line to explain how I fixed it.

zenhack replied:
@SahilTikale, would still like this comment removed.

SahilTikale replied:
removed it.

zenhack commented on tests/unit/dev_support.py:
```diff
@@ -54,13 +54,17 @@ def _wet(func):
 # Actual test cases:
-def test_dry_function(): _dry(_function)
+def test_dry_function():
+    _dry(_function)
```
Can you explain why you changed these? Not really a problem, just doesn't seem to have anything to do with this PR, and it doesn't really seem like an improvement (though not really worse either).

SahilTikale replied:
I did not make this changes. I do not know how this got into my PR.

naved001 replied:
Probably @meng-sun / @gfaline changed this while trying to fix pep8 errors. That's my best guess.

zenhack commented on haas/client/auth.py:
```diff
@@ -0,0 +1,32 @@
+import requests
```
I don't like the way you've structured the auth stuff at all. Here's what I would do:

1. Take the HTTPClient classes as they are on master (i.e. without any of the modifications you've done), and move them to the client library.
2. Rather than having the session parameter to the Client class, just accept an HTTPClient.

Then you can just delete this file.

The auth stuff is already done and in a usable form on master; you should just use it as is.

SahilTikale replied:
I tried refactoring it the way you suggested but I think this requires a little more discussion. 
Global variables `http_client` and `C` will have to coexist until all calls have migrated to the client library. 

Lets do it when we meet this week. Will mail you with suitable day and time.

zenhack commented on haas/cli.py:
```diff
@@ -180,17 +186,22 @@ def setup_http_client():
     This may be extended with other backends in the future.
     """
```
The docstring for this function will need to be updated to reflect the changes this patch makes. I would wait until after you've sorted out the auth stuff I mentioned though, as that's likely to involve further changes to exactly what this does.

henn commented on haas/client/switch.py:
```diff
@@ -0,0 +1,59 @@
+        """Registers a switch with name <switch> and
+        model <subtype> , and relevant arguments  in <*args>
+        """
+#       It is assumed that the HIL administrator is aware of
```
We should do something like `raise NotImplemented` if the function doesn't do what it claims

SahilTikale replied:
done.

henn commented on haas/client/node.py:
```diff
@@ -0,0 +1,94 @@
+#       FIXME: In future obm_types should be dynamically fetched
+#        from haas.cfg, need a new api call for querying available
+#        and currently active drivers for HIL
+#        Not implemented yet.
```
We should do something like `raise NotImplemented` if the function doesn't do what it claims

zenhack commented:
Quoting Sahil Tikale (2017-02-28 18:32:42)

>    I did not understand your comment about if these are mandatory just
>    don't supply the defaults.
>    I am checking if they are explicitly None and raises error in that
>    case.

Instead of doing:

    def __init__(self, endpoint=None, sess=None):
        ...
        if None not in [endpoint, session]:
            ...

Just do:

    def __init__(self, endpoint, sess):
        ...

Adding `=None` to the arguments doesn't make any sense, because the user
*needs* to supply a value. If you leave that off, they'll get a
TypeError when they call the function without those, and you won't have
to check explicitly in the body of the function.

Is that more clear?

zenhack commented:
Quoting Sahil Tikale (2017-02-28 18:37:52)

> >    Should I write Not implemented in the doc string or as a comment.

Docstring would be better, as it shows up in tools like pydoc. This is
generally true of anything the caller of the function should know.
Comments should be for people looking at the implementation.

SahilTikale commented:
@zenhack made all the changes you mentioned.
Let me know if you need anything else.

zenhack commented:
@SahilTikale, there were a couple things you missed, but we're getting very close.

zenhack commented on haas/cli.py:
```diff
@@ -695,7 +695,11 @@ def show_network(network):
     FIXME: Recursion should be implemented to the output.
+    The output of show_node is a dictionary that can be list of list, having
+    multiple nics and networks. More meta data about node could be shown
+    via this call. So this is a note for developers that will work on CLI.
```
Could you move this to a comment just below? the docstring gets displayed to the user of cli tool, which isn't really appropriate for this information.

SahilTikale replied:
done.

zenhack commented on tests/integration/cli.py:
```diff
@@ -18,21 +18,18 @@ def test_add_list_delete_projects():
     subprocess.check_call(['haas', 'project_delete', PROJECT1])
     output = subprocess.check_output(['haas', 'list_projects'])
     projects = output.split(" ")
-    projects.remove('\n')  # remove \n from list
```
Maybe I wasn't clear; it was just the *comment* that was superfluous, not the statement.

SahilTikale replied:
With the changes I have made to how the output is presented, the said statement was no longer required. Having that statement was failing the tests, removing it fixed it. I confirmed it with @naved001 before committing the changes.

SahilTikale replied:
While interfacing the client library with CLI, I had changed the way calls output results. 
because of that the said statement is no longer required. Infact the integration tests were failing when I made the suggested changes, because of the above reason. Just to make sure I confirmed with @naved001 before committing the changes.

zenhack replied:
Fair enough.

SahilTikale commented:
@zenhack let me know if you require any more changes.

zenhack commented:
Quoting Sahil Tikale (2017-03-27 08:40:40)

> @zenhack let me know if you require any more changes.

Will do -- probably will have a chance to take a look at this late
Tuesday (tomorrow).

SahilTikale commented on haas/client/base.py:
```diff
@@ -0,0 +1,40 @@
+        sess: depending on the authentication backend (db vs keystone) the
+        parameters required to make up the session vary.
+        user: username as which you wish to connect to HaaS
+        Currently all this information is fetched from the user's environment.
```
This has changed. Now the client library takes a complete http client object not just a session authentication object.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+    sess = requests.Session()
+    sess.auth = (username, password)
+
+    # Adding nodes, node-01 - node-06
```
This comment is wrong; the code is creating nodes `node-01` through `node-09`.

SahilTikale replied:
Next commit fixes it.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+            url_project + 'proj-01' + '/connect_node',
+            data=json.dumps({'node': 'node-01'})
+            )
+    # Adding nodes 2, 4 to proj-02
```
A comment like this jumps out and says "for loop."

SahilTikale replied:
Since each project was allocated different number of nodes, a for loop was not feasible. But I get your point. Hence wrote a function to assign variable number of nodes to a given project. Next commit will have that.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+
+    # Adding nodes, node-01 - node-06
+    url_node = 'http://127.0.0.1:8000/node/'
+    api_nodename = 'http://schema.massopencloud.org/haas/v0/obm/'
```
You don't look to be using any obm types other than ipmi, and you only use this variable in one place. Why not just put the literal ipmi url in the dict below? The variable name also makes no sense.

zenhack replied:
...or just use OBM_TYPE_IPMI, which you define above.

SahilTikale replied:
I defined the variable `api_nodename` because, using it everytime would exceed the character width of 79 and violate pep8. To reduce unnecessary lines of code I defined it as a variable and used it. Later I made it a for loop and hence it was used only once. I have renamed the variable name from `api_nodename` to `ipmi`. This is better than using a string literal for future tests if we have a case of registering nodes with other obm types. Next commit has the changes.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+                )
+
+
+# -- SETUP --
```
Don't know why this comment is here?

SahilTikale replied:
While developing tests It helped to jump to different sections of file between, config, populate and setup. Removed it.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+    proc2 = run_server(['haas', 'serve_networks'])
+    time.sleep(1)
+    populate_server()
+    print("coming from create_setup")
```
Don't leave debugging print statements in the source. Same thing in fin().

SahilTikale replied:
Next commit removes it.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+            C.network.revoke_access('proj-02', 'newnet03')
+
+
+# End of tests ##
```
This is superfluous.

SahilTikale replied:
next commit removes it.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+#       5. tears down the setup in a clean fashion.
+
+
+# pytest.fixture(scope="module")
```
I take it make_config was a fixture at one time. I think this actually makes more sense; we should make it a fixture, and allso add the call to cleanup as a finalizer (It wasn't obvious at the call site what cleanup was doing).

SahilTikale replied:
`make_config` is not a fixture. `create_setup` is a fixture that calls both make_config and `fin` calls `cleanup`.

SahilTikale replied:
removed that comment.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+class Test_ClientBase:
+    """ When the username, password is not defined
+    It should raise a LookupError
+    """
```
This docstring doesn't seem to relate to any of the tests in this class?

SahilTikale replied:
Next commit fixes the docstring.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+        try:
+            x = ClientBase()
+        except TypeError:
+            assert True
```
Use `pytest.raises`

SahilTikale replied:
next commit fixes it.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+#    def test_correct_init(self):
+#        x = ClientBase(ep, 'some_base64_string')
+#        assert x.endpoint == "http://127.0.0.1:8000"
+#        assert x.auth == "some_base64_string"
```
Don't leave code commented out. This looks like it was testing something that's obviated by the change to take an http client, so we should just delete it.

For reference, if you have a test that is failing, fixing it is preferred, but if there's some reason you're not going to (e.g. the test is for a discovered bug that hasn't been sorted out yet), you can use `pytest.mark.xfail`.

SahilTikale replied:
it does not show in my `client.py` I may have removed it already.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+# def test_node_detach_network(self):
+# C.node.connect_network('node-04', 'eth0', 'net-04', 'vlan/native')
+# result = C.node.detach_network('node-04', 'eth0', 'net-04')
+# assert result is None
```
Same thing as above; don't leave code commented out. Can you explain to me what was happening here?

SahilTikale replied:
I spent considerable time debugging this. since networking actions are asynchronous actions, there is no reliable way to check node_detachment. Trying to detach nodes from network with real switch drivers hangs because the server timeouts waiting for response from real switch. With Mock switch drivers, there is no feedback when the detachment happens. Hence the test is unreliable. I will remove it for now. I updated this comments in the file and added `pytest.mark.xfail`

zenhack commented on haas/cli.py:
```diff
@@ -215,17 +142,27 @@ def setup_http_client():
 
 
 class FailedAPICallException(Exception):
     pass
```
This shouldn't be here anymore; you've moved it to the lib.

SahilTikale replied:
This is right. Based on which http client is used (http, keystone, no authentication) this statement initializes the client library that subsequent calls of cli make use of.

zenhack replied:
Not sure what you mean? I'm talking about the definition of FaileAPICallException -- rather than defining it again here, just import it from haas.client.base.

SahilTikale replied:
I misunderstood your comment last time was pointing at the code that initializes the client library. 
Removed the definition and imported same from the client library. Next commit would have the changes.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+        """ test for erronous reconnecting node to project. """
+        C.project.create('proj-05')
+        C.project.connect('proj-05', 'node-07')
+        with pytest.raises(Exception):
```
This should be FailedAPICallException; same wtih the rest of them.

SahilTikale replied:
I did not understand. How is this test different from others ?

zenhack replied:
By "same with the rest of them," I mean the other tests need to be changed in the same way, not that it should be changed to what the other tests are doing.

SahilTikale replied:
next commit fixes it.

zenhack commented on tests/unit/client.py:
```diff
@@ -0,0 +1,659 @@
+        result01 = C.project.nodes_in('proj-01')
+        result02 = C.project.nodes_in('proj-02')
+        assert result01 == [u'node-01']
+        assert result02 == [u'node-02', u'node-04']
```
Most of these tests would be much less noisy if you combined the call and the assert, e.g.

```python
assert C.project.nodes_in('proj-01') == [u'node-01']
assert C.project.nodes_in('proj-02') == [u'node-02', u'node-04']
```

SahilTikale replied:
next commit implements these changes for all tests wherever applicable.

zenhack commented on haas/client/node.py:
```diff
@@ -0,0 +1,99 @@
+        obm_api = "http://schema.massopencloud.org/haas/v0/obm/"
+        obm_types = ["ipmi", "mock"]
+#       FIXME: In future obm_types should be dynamically fetched
+#        from haas.cfg, need a new api call for querying available
```
Get rid of "from haas.cfg"; the API call makes more sense as you describe, but wherever we get it from it's not going to be the server's config file.

SahilTikale replied:
next commit fixes it.

SahilTikale replied:
next commit removes that string.

zenhack commented on haas/cli.py:
```diff
@@ -806,15 +739,13 @@ def show_console(node):
     """Start logging console output from <node>"""
-    url = object_url('node', node, 'console')
-    do_put(url)
+    q = check_clientlib_response(lambda: C.node.start_console(node))
```
Don't store the return value in a variable unless you're actually going to use it. same thing below.

SahilTikale replied:
agreed. next commit fixes it.

zenhack commented on haas/cli.py:
```diff
@@ -215,17 +142,27 @@ def setup_http_client():
+    try:
+        return fun()
+    except Exception as e:
+        sys.stderr.write('Error: %s\n' % e.message)
```
By catching the exception you're causing the function to just return None, so on failure callers are going to plow ahead and probably raises some confusing exception. Probably the thing to do is call sys.exit instead of sys.stderr.write.

SahilTikale replied:
next commit fixes it.

zenhack commented:
@SahilTikale, more coments. I think I've done a deep enough dive now that you shouldn't get anything more from me besides "missed this one"

zenhack commented on tests/unit/client.py:
```diff
@@ -335,35 +301,37 @@ def test_nosuch_node_add_nic(self):
+# When using a real switch driver for network, haas network server
+# times out waiting for response from switch while using mock switch
+# driver, the networking_action queue takes longer to execute the request
+# the testing server kills the servers when other tests are completed.
```
This paragraph scans very poorly. I read it 7 or 8 times and still wasn't entirely sure what it meant. Then I went back and looked at your comment, which is more understandable.

You're missing some punctuation, but that's not everything that makes it hard to understand.

The explanation you gave in a comment

zenhack commented:
Just a few more things that need attention.

SahilTikale commented:
will make the necessary changes by today. Are you available to chat in the evening or night for sometime ? Wanted to some inputs on the two of your comments. I am still having difficulty understanding what you want me to do.

zenhack commented:
Quoting Sahil Tikale (2017-04-03 17:15:51)

>    will make the necessary changes by today. Are you available to chat in
>    the evening or night for sometime ? Wanted to some inputs on the two of
>    your comments. I am still having difficulty understanding what you want
>    me to do.

I'm in the #moc channel on Freenode; will probably be there for most of
the evening. Drop me a line.

SahilTikale commented:
@zenhack fixed the test. I think I have addressed all the issues from you so far.
Let me know if anything else is needed.

zenhack commented on tests/unit/client.py:
```diff
@@ -308,31 +308,34 @@ def test_remove_duplicate_nic(self):
-    @pytest.mark.xfail
     def test_node_detach_network(self):
         C.node.connect_network('node-04', 'eth0', 'net-04', 'vlan/native')
+        time.sleep(1)
```
You've got a race condition here. I'm fine leaving this as-is *if* you:

1. Hit the re-run button on travis several times to make sure this is a sufficient delay.
2. Add a comment pointing out that there is a race condition, which could theoretically crop up.

Same thing goes for the call to sleep in the next function.

zenhack commented on tests/unit/client.py:
```diff
@@ -326,12 +326,12 @@ def test_node_connect_network_error(self):
     def test_node_detach_network_error(self):
         C.node.connect_network('node-04', 'eth0', 'net-04', 'vlan/native')
-        time.sleep(1)
+        time.sleep(2)  # To avoid a potential race condition
```
Clarification: Nothing you pass to sleep is going to eliminate the race condition. What I'm looking for is (1) a value that's big enough that failures will be rare, and (2) a comment that calls attention to the fact that there *is* a race condition. This seems to imply that the race condition is elimnated by the call to sleep, which is not true.

zenhack commented on tests/unit/client.py:
```diff
@@ -236,6 +240,21 @@ def populate_server():
 
 
+# Avoid race condition while creating, deleting networks
+def avoid_network_race_condition():
```
1. Can't we do this using the SQLAlchemy APIs? I'd prefer to limit the number of ways we write code to access a db.
2. Once we have a show_networkingaction or such in the API, we should probably replace this with calls to that; would you put a note in to that effect?

SahilTikale replied:
done using SQLAlchemy APIs.

zenhack commented on tests/unit/client.py:
```diff
@@ -236,6 +240,21 @@ def populate_server():
+    conn = sqlite3.connect(db_file)
+    cur = conn.cursor()
+    while True:
+        que = len(cur.execute('select * from networking_action;').fetchall())
```
You could just do `SELECT count(*) ...` and omit the call to `len`. With SQLAlchemy there's a `count` method that we use elsewhere in the source.

SahilTikale replied:
`len` omitted. done using SQLAlchemy.

zenhack commented on tests/unit/client.py:
```diff
@@ -236,6 +240,21 @@ def populate_server():
+    while True:
+        que = len(cur.execute('select * from networking_action;').fetchall())
+        if que > 0:
+            time.sleep(1)
```
I think we should still put some upper bound on how long we're willing to wait -- maybe a minute or two. Otherwise travis will run for a *very* long time without notifying us that anything is wrong. I've had jobs just busy looping for 20 minutes before.

Doing an exponential backoff would be the "standard" approach, but I'm not picky on that point.

SahilTikale replied:
put a time out of 10 sec

SahilTikale commented:
hi @zenhack as per your suggestion I wrote a new function `avoid_race_sqlalchemy` using sqlalchemy. Compared to the previous function `avoid_network_race_condition`  I find using sqlalchemy requires more imports (extra loc for accomplishing same thing) and a dependency on db.model since it requires mapping between tables and objects. I still feel for the sake of simplicity and keeping client library as much decoupled as possible from rest of the HIL code, we should choose the `avoid_network_race_condition`  implementation over  `avoid_race_sqlalchemy`.

SahilTikale commented:
Also I have added a timeout of 10 sec so as to avoid longer wait times. I ran the test multiple times and found 7 seconds to be adequate time for the test to complete successfully. A timeout of 10 should be sufficient.

zenhack commented on tests/unit/client.py:
```diff
@@ -247,10 +247,41 @@ def avoid_network_race_condition():
+        que = session.query(
+                func.count('*')).select_from(
+                        NetworkingAction
+                        ).scalar()
```
Could just be `session.query(NetworkingAction).count()`

SahilTikale replied:
Used that. assigned it a variable condition so that I can use the value in the variable twice, once for testing the if condition and to return it so as to determine the time out.

zenhack commented on tests/unit/client.py:
```diff
@@ -247,10 +247,41 @@ def avoid_network_race_condition():
+        if que > 0:
+            time.sleep(1)
+            timeout = timeout + 1
+            if timeout > 10:  # breakout if exceeds 10 secs.
```
You should fail the test with some clear error message if this happens.

SahilTikale replied:
clear message included.

zenhack commented on tests/unit/client.py:
```diff
@@ -247,10 +247,41 @@ def avoid_network_race_condition():
+    session = Session()
+    timeout = 0
+
+    while True:
```
Rather than setting a variable and incrementing it each time through the loop, just do `for _ in range(10):`. Add an assert *after* the loop that covers the case where it times out, and do a `return` inside the loop when the operation finishes.

SahilTikale replied:
I will do the `for in range(10):` instead of a while loop. about the assert after the loop, I did that but then revised it to simply breaking out of the function when timeout exceeds, because that way the test fails anyways and with an appropriate message like 
`E           FailedAPICallException: A networking operation is already active on the nic.
`
Full output of such a failure test is given below. Let me know if you still want me to change the code to fail deliberately after the timeout period. ?

```
tests/unit/client.py F

==================================================== FAILURES =====================================================
____________________________________ Test_node.test_node_detach_network_error _____________________________________
self = <tests.unit.client.Test_node instance at 0x7f07914a1d88>

    def test_node_detach_network_error(self):
        C.node.connect_network('node-04', 'eth0', 'net-04', 'vlan/native')
    #    avoid_network_race_condition()
>       C.node.detach_network('node-04', 'eth0', 'net-04')

/home/sahil/git/HIL/tests/unit/client.py:385: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/sahil/git/HIL/haas/client/node.py:84: in detach_network
    self.httpClient.request('POST', url, data=payload)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <haas.client.node.Node object at 0x7f079140a490>, response = <Response [409]>

    def check_response(self, response):
        if response.ok:
            if response.request.method == 'GET':
                return response.json()
            else:  # For methods PUT, POST, DELETE
                return
        else:
            e = response.json()
>           raise FailedAPICallException(e['msg'])
E           FailedAPICallException: A networking operation is already active on the nic.

/home/sahil/git/HIL/haas/client/base.py:49: FailedAPICallException
============================================ 1 failed in 5.57 seconds =============================================
```

zenhack replied:
Yeah, I think it should specifically say something about a timeout.

SahilTikale replied:
Mentioned specifically about the timeout when it fails due to the race condition time out.

zenhack commented on tests/unit/client.py:
```diff
@@ -247,10 +247,41 @@ def avoid_network_race_condition():
+    uri = 'sqlite:///'+db_dir+'/haas.db'
+    engine = create_engine(uri)
+    Session = sessionmaker(bind=engine)
+    session = Session()
```
You could cut this down a bit by just using db.session, and a `with app.app_context():` None of the session setup, besides setting the URI in the config. It also gets rid of most of the imports above.

SahilTikale replied:
Done on similar lines. reduced the number of imports.

zenhack commented on tests/unit/client.py:
```diff
@@ -247,10 +247,41 @@ def avoid_network_race_condition():
+    from haas.model import Node, Project, NetworkingAction
+    from sqlalchemy import func
+    from sqlalchemy import create_engine
+    from sqlalchemy.orm import sessionmaker
```
imports at the top of the file.

SahilTikale replied:
I wanted to keep the imports local, so that we can remove them when the `show_networking_action` is implemented. I will update the function with the relevant comment. What do you think ? Leave it local or not.

zenhack replied:
Always put imports at the top of the file, unless there's a *strong* reason not to (e.g. we can't do this with extensions in the tests, since otherwise the tests pollute each other). If you're importing stuff inside of a function, it means you're finding out about related errors later, which is generally a bad thing. It also makes it harder to understand what code you're actually pulling in, and when.

There are plenty of linters out there that can tell us about unused imports. One of these days we should rig one up to the CI and enforce things that way. That will be a much more reliable way of making sure we remember to remove stuff.

SahilTikale replied:
OK, Agreed. Next commit fixes it.

zenhack commented:
A 10 second timeout is fine with me.

I made a number of comments suggesting ways to make the code more clear & concise.

Re: pulling in server code, given that what it's doing is poking at a database that only the server is supposed to have access to, I don't think it's a problem. If it  were doing something that should be possible client-side, that would be a huge red flag. Once we have show_networking_action or such, we should use that here instead.

shwsun commented on haas/cli.py:
```diff
@@ -168,6 +86,8 @@ def wrapped(*args, **kwargs):
 def setup_http_client():
     """Set `http_client` to a valid instance of `HTTPClient`
 
```
we should get rid of this line

SahilTikale replied:
This is a valid doc string. why should we get rid of it ?

shwsun commented on haas/cli.py:
```diff
@@ -235,9 +169,11 @@ def check_status_code(response):
+
 def object_url(*args):
     # Prefer an environmental variable for getting the endpoint if available.
     url = os.environ.get('HAAS_ENDPOINT')
```
There are blank lines between documentation and functions that we should get rid of. 

About the `object_url` and other functions, I think we should move them to the end of `cli.py` and specify that they will be soon replaced by xxx. It is just a suggestion, what you you think @SahilTikale ?

SahilTikale replied:
that and subsequent four functions were there prior to this PR. I do not see the point of changing their location if that is not going to add any value to the code. I have specified that they will soon be deleted before `def object_url` and after `def do_delete`

shwsun commented on haas/cli.py:
```diff
@@ -666,43 +584,40 @@ def switch_register(switch, subtype, *args):
-    do_post(url, data={'node': node, 'nic': nic})
+    check_clientlib_response(
+            lambda: C.port.connect_nic(switch, port, node, nic)
+            )
```
I am a bit bothered by the trailing one here, can you run some thing like yapf formatter to figure out how to indent it? Only suggestions

SahilTikale replied:
@shwsun as we discussed, I am keeping the code on single line as long as it is within 79 characters. If it exceeds it break it like above at the brackets. I test all files for any pep8 violations. This style of breaking the line of code is acceptable by those standards.

shwsun commented on haas/cli.py:
```diff
@@ -725,54 +640,69 @@ def list_nodes(is_free):
+#    The output of show_node is a dictionary that can be list of list, having
+#    multiple nics and networks. More metadata about node could be shown
+#    via this call. Suggestion to future developers of CLI to use
+#    recursion in the call for output of such metadata.
```
indent or move them to `rest_api.md`

SahilTikale replied:
Not moving anything to `rest_api.md` as it is not relevant.

shwsun commented on haas/client/base.py:
```diff
@@ -0,0 +1,49 @@
+""" This module implements the HaaS client library. """
```
Could you add the software copyright part here?

shwsun commented on haas/client/client.py:
```diff
@@ -0,0 +1,108 @@
+from haas.client.base import ClientBase
```
copyright here

shwsun commented on haas/client/node.py:
```diff
@@ -0,0 +1,99 @@
+
+    def show(self, node_name):
+        """Shows attributes of a given node """
+
```
blank line we should get rid of

SahilTikale replied:
next commit fixes it.

shwsun commented on haas/client/node.py:
```diff
@@ -0,0 +1,99 @@
+#       available OBM driver and its corresponding arguments.
+#       We assume that the HIL administrator is aware as to which
+#       Node requires which OBM, and knows arguments required
+#       for successful node registration.
```
indent or move to rest_api

SahilTikale replied:
Not moving it anything to `rest_api` as discussed before.

shwsun commented on haas/client/node.py:
```diff
@@ -0,0 +1,99 @@
+        obm_types = ["ipmi", "mock"]
+#       FIXME: In future obm_types should be dynamically fetched.
+#        We need a new api call for querying available
+#        and currently active drivers for HIL
```
indent

SahilTikale replied:
next commit fixes it.

shwsun commented on haas/client/user.py:
```diff
@@ -0,0 +1,52 @@
+
+class User(ClientBase):
+    """Consists of calls to query and
+
```
blank line

zenhack commented:
@shwsun, `rest_api.md` documents the HTTP Api; it shouldn't be  discussing code structure *at all,* and nothing in this PR changes the API, so nothing in that file ought to change.

naved001 commented:
@shwsun you should edit your comment to hide login details in the error log. this page is public.

shwsun commented:
@naved001 good catch, thank you!

zenhack commented:
@naved001, @shwsun, I didn't catch what it was that got leaked, and I'm sure this goes without saying,  but if there were credentials in there, *change them*.

SahilTikale commented on tests/unit/client.py:
```diff
@@ -0,0 +1,635 @@
+def make_config():
+    """ This function creates haas.cfg with desired options
+    and writes to a temporary directory.
+    It returns a tuple where (tmpdir, cwd) = ('location of haas.cfg', 'pwdd')
```
Looks like your configuration has some problems.
Is the server running on port 5000 or port 5050

`DEBUG:requests.packages.urllib3.connectionpool:http://127.0.0.1:5050`
or
`Error: HTTPConnectionPool(host='127.0.0.1', port=5000): Max retries exceeded with url...`

SahilTikale replied:
Create an init_client file with correct values for following parameters.
```
export HAAS_ENDPOINT=http://<ip address + port no. where haas server is running>
export HAAS_USERNAME=<authorized username>
export HAAS_PASSWORD=<password for that username>
```
source this init_client file in your environment as we normally do.

SahilTikale replied:
thanks for catching the typo, next commit fixes it.

SahilTikale commented:
@zenhack, I believe I have addressed all your comments, let me know if anything else is needed.

shwsun commented:
> @naved001, @shwsun, I didn't catch what it was that got leaked, and I'm sure this goes without saying, but if there were credentials in there, change them.

@zenhack The thing that naved pointed out were some ip addr/passwd that I included previously in my comment, those sensitive info were already gotten rid of.

zenhack commented on tests/unit/client.py:
```diff
@@ -248,21 +250,18 @@ def avoid_network_race_condition():
         else:
-            break
+            return que
+    return timeout
```
I would just raise an exception here (of a newly-declared type, which denotes a timeout), and get rid of the if/else checks elsewhere.

zenhack commented:
@SahilTikale, travis is failing. Given that it's just postgres I'm wondering whether 10 seconds is enough to be reliable, but I'm not totally sure what's going on.

Other than that I'm happy; get the tests passing reliably again and I'll approve.

SahilTikale commented:
@zenhack  passes all the tests. 
Time out is not a problem. sometimes it is not able to spin up the server on the designated port and that causes failures.

SahilTikale commented:
@shwsun only your approval is pending.

Let me know if you need me to address any issues.

shwsun commented:
@SahilTikale I see, could you be around sometime tmr? I still have some questions on deployment

naved001 commented:
@SahilTikale The tests are failing again. Probably due to the race condition thing you mentioned, it might even pass if we re-run this.
```
socket.error: [Errno 98] Address already in use
```
But I think we should *reliably* fix that test here, otherwise we would have a tough time with CI failing on future PRs.

zenhack commented:
@SahilTikale, I agree with @naved001; the tests need to pass reliably.

SahilTikale commented:
@naved001  the race condition was a different problem. There is a function to solve that. 
This is a different issue of not being able to find the port no to spin up the server and it is not that frequent. 

@zenhack  there is nothing to worry.

SahilTikale commented:
@shwsun I have addressed the issue we were discussing today.
Let me know if anything else is needed.

SahilTikale commented:
@shwsun I would like to get this PR merged if there is nothing else that you would want me to change. 

Let me know if you need anything else.

shwsun commented:
I dont think I have any other more concerns. +1

SahilTikale commented:
@zenhack would you merge it or should I do it.
